lib/posix/pthread_mutex.c

```diff
@@ -9,16 +9,214 @@
 #include <wait_q.h>
 #include <posix/pthread.h>
 
-int pthread_mutex_trylock(pthread_mutex_t *m)
+/* Memory pool for pthread space */
+K_MEM_POOL_DEFINE(posix_mutex_sem, sizeof(struct k_sem),
+		  sizeof(struct k_sem), 32, 4);
+
+/*
+ *  Default mutex attrs.
+ */
+static const pthread_mutexattr_t def_attr = {
+	.type = PTHREAD_MUTEX_DEFAULT,
+	.protocol = PTHREAD_PRIO_NONE,
+};
+
+static int mutex_recursive_take(pthread_mutex_t *m, int timeout)
+{
+	if (m->lock_count == 0) {
+		m->owner = pthread_self();
+	}
+
+	if (m->owner == pthread_self()) {
+		m->lock_count++;
+		if (m->lock_count > MUTEX_MAX_REC_LOCK) {
+			return EAGAIN;
+		}
+		return k_sem_take(m->sem, timeout);
+	}
+
+	return EPERM;
+}
+
+static int acquire_mutex(pthread_mutex_t *m, int timeout)
 {
-	int key = irq_lock(), ret = EBUSY;
+	pthread_t this_thread;
+
+	this_thread = pthread_self();
+	int rc = 0;
+
+	if (m->owner == this_thread) {
+		if (m->type == PTHREAD_MUTEX_ERRORCHECK) {
+			return EDEADLK;
+		}
+	}
+
+	if (m->type == PTHREAD_MUTEX_RECURSIVE) {
+		rc = mutex_recursive_take(m, timeout);
+	} else {
+		m->lock_count = 1;
+		rc = k_sem_take(m->sem, timeout);
```

Comment:
acquire_mutex() is called while the IRQs are locked.  If one passes a timeout of K_FOREVER, and the semaphore is taken, the system (and not the thread!) will freeze forever.